Compiling on ARM 64bit

Hi all.

I'm building `Pioneer` on `ARM-64bit` architectures for Fedora. The release `20160907` is successfully compiled: [log](https://kojipkgs.fedoraproject.org//packages/pioneer/20160907/1.fc26/data/logs/aarch64/build.log) / [list software used](https://kojipkgs.fedoraproject.org//packages/pioneer/20160907/1.fc26/data/logs/aarch64/root.log)

Latest release `20161022` is failing on `ARM64` but i can't recognize the reason. Can you take a look please?
[Build log](https://kojipkgs.fedoraproject.org//work/tasks/8451/16188451/build.log)
[Dependencies log](https://kojipkgs.fedoraproject.org//work/tasks/8451/16188451/root.log)

Thanks.

To @fluffyfreak:
Perhaps, this issue is not right place to discuss about mentioned patches, but since you noted about them here:

Some descriptions:
84amd64: FreeBSD 8.4 amd64, GNU/GCC 4.8.2 with GCC 4.2 includes in base (which is [EOL](https://www.freebsd.org/security/unsupported.html) currently).
93amd64: FreeBSD 9.3 amd64, GNU/GCC 4.8.5 (with possible [EOL](https://www.freebsd.org/security/security.html#sup) on December 31, 2016).
102amd64: FreeBSD 10.2 amd64, LLVM/Clang 3.4.1 (with possible [EOL](https://www.freebsd.org/security/security.html#sup) on December 31, 2016).

1) https://github.com/freebsd/freebsd-ports/blob/f81a8530f4df1040b94dcac6c466b8a1202a1dfb/games/pioneer/files/patch-src_GameConfig.h
84amd64: this was proposed in [PR 193213](https://bugs.freebsd.org/193213).
93amd64: ok.
102amd64: ok.

2) https://github.com/freebsd/freebsd-ports/blob/f81a8530f4df1040b94dcac6c466b8a1202a1dfb/games/pioneer/files/patch-contrib_PicoDDS_PicoDDS.cpp
93amd64: ok.
102amd64: with error message:
```shell
PicoDDS.cpp:423:23: error: use of undeclared identifier 'malloc'; did you mean 'valloc'?
        char* pData = (char*)malloc(sizeof(char)*size);
                             ^~~~~~
                             valloc
/usr/include/unistd.h:583:7: note: 'valloc' declared here
void    *valloc(size_t);                        /* obsoleted by malloc() */
         ^
PicoDDS.cpp:430:3: error: use of undeclared identifier 'free'
                free(pData);
                ^
2 errors generated.
```
this was proposed in [PR 193213](https://bugs.freebsd.org/193213).

3) https://github.com/freebsd/freebsd-ports/blob/ad301a025fc7d38f935abc6b8068a182ebedc276/games/pioneer/Makefile#L89
or
`sed -e 's/std:://g' src/Orbit.cpp`
or
```diff
--- src/Orbit.cpp.orig
+++ src/Orbit.cpp
@@ -145,7 +145,7 @@ double Orbit::OrbitalTimeAtPos(const vec
 	vector3d approx3dPos = (matrixInv * pos - vector3d(c, 0., 0.)).Normalized();
 
 	double cos_v = -vector3d(1., 0., 0.).Dot(approx3dPos);
-	double sin_v = std::copysign(vector3d(1., 0., 0.).Cross(approx3dPos).Length(), approx3dPos.y);
+	double sin_v = copysign(vector3d(1., 0., 0.).Cross(approx3dPos).Length(), approx3dPos.y);
 
 	double cos_E = (cos_v + m_eccentricity) / (1. + m_eccentricity * cos_v);
 	double E;
@@ -159,7 +159,7 @@ double Orbit::OrbitalTimeAtPos(const vec
 	}
 	else
 	{
-		E = std::acosh(cos_E);
+		E = acosh(cos_E);
 		if(sin_v < 0)
 			E *= -1.;
 		meanAnomaly = E - m_eccentricity * std::sinh(E);

```
93amd64: with error message:
```shell
Orbit.cpp: In member function 'double Orbit::OrbitalTimeAtPos(const vector3d&, double) const':
Orbit.cpp:148:17: error: 'copysign' is not a member of 'std'
  double sin_v = std::copysign(vector3d(1., 0., 0.).Cross(approx3dPos).Length(), approx3dPos.y);
                 ^
Orbit.cpp:148:17: note: suggested alternative:
In file included from /usr/local/include/SDL2/SDL_stdinc.h:81:0,
                 from /usr/local/include/SDL2/SDL_main.h:25,
                 from /usr/local/include/SDL2/SDL.h:32,
                 from libs.h:12,
                 from Orbit.h:7,
                 from Orbit.cpp:4:
/usr/include/math.h:268:8: note:   'copysign'
 double copysign(double, double) __pure2;
        ^
Orbit.cpp:162:7: error: 'acosh' is not a member of 'std'
   E = std::acosh(cos_E);
       ^
Orbit.cpp:162:7: note: suggested alternative:
In file included from /usr/local/include/SDL2/SDL_stdinc.h:81:0,
                 from /usr/local/include/SDL2/SDL_main.h:25,
                 from /usr/local/include/SDL2/SDL.h:32,
                 from libs.h:12,
                 from Orbit.h:7,
                 from Orbit.cpp:4:
/usr/include/math.h:222:8: note:   'acosh'
 double acosh(double);
        ^
  CXX      Pi.o
```
this was proposed in [PR 203427](https://bugs.freebsd.org/203427).
102amd64: ok.

4.1)
sed -e 's/std::to_string/to_string/g' src/graphics/opengl/GenGasGiantColourMaterial.cpp
or
```diff
--- src/graphics/opengl/GenGasGiantColourMaterial.cpp.orig
+++ src/graphics/opengl/GenGasGiantColourMaterial.cpp
@@ -48,7 +48,7 @@
 	}
 	// extract the top 16-bits to get how many octaves we can use
 	const Uint32 octaves = (desc.quality & 0xFFFF0000) >> 16;
-	ss << "#define FBM_OCTAVES " << std::to_string(octaves) << std::endl;
+	ss << "#define FBM_OCTAVES " << to_string(octaves) << std::endl;
 	
 	// No lights
 	ss << "#define NUM_LIGHTS 0\n";
```
93amd64: with error message:
```shell
GenGasGiantColourMaterial.cpp: In constructor 'Graphics::OGL::GenGasGiantColourProgram::GenGasGiantColourProgram(const Graphics::MaterialDescriptor&)':
GenGasGiantColourMaterial.cpp:51:34: error: 'to_string' is not a member of 'std'
  ss << "#define FBM_OCTAVES " << std::to_string(octaves) << std::endl;
                                  ^
GenGasGiantColourMaterial.cpp:51:34: note: suggested alternative:
In file included from GenGasGiantColourMaterial.cpp:9:0:
./../../StringF.h:145:20: note:   'to_string'
 inline std::string to_string(const T& value) {
                    ^
```
was found while testing 20161022 release on FreeBSD 9.3 amd64.
102amd64: ok.

4.2) https://github.com/freebsd/freebsd-ports/blob/ad301a025fc7d38f935abc6b8068a182ebedc276/games/pioneer/Makefile#L91
`sed -e 's/std::to_string/to_string/g' src/graphics/opengl/RendererGL.cpp`
or
```diff
--- src/graphics/opengl/RendererGL.cpp.orig
+++ src/graphics/opengl/RendererGL.cpp
@@ -317,7 +317,7 @@ void RendererOGL::CheckErrors(const char
 		// now build info string
 		std::stringstream ss;
 		if(func) {
-			ss << "In function " << std::string(func) << "\nOn line " << std::to_string(line) << "\n";
+			ss << "In function " << std::string(func) << "\nOn line " << to_string(line) << "\n";
 		}
 		ss << "OpenGL error(s) during frame:\n";
 		while (err != GL_NO_ERROR) {

```
93amd64: with error message:
```shell
RendererGL.cpp: In static member function 'static void Graphics::RendererOGL::CheckErrors(const char*, int)':
RendererGL.cpp:320:65: error: 'to_string' is not a member of 'std'
    ss << "In function " << std::string(func) << "\nOn line " << std::to_string(line) << "\n";
                                                                 ^
RendererGL.cpp:320:65: note: suggested alternative:
In file included from RendererGL.cpp:9:0:
./../../StringF.h:145:20: note:   'to_string'
 inline std::string to_string(const T& value) {
                    ^
```
this was proposed in [PR 209357](https://bugs.freebsd.org/209357).
102amd64: ok.

--- a/contrib/profiler/Profiler.cpp
+++ b/contrib/profiler/Profiler.cpp
@@ -24,6 +24,7 @@
 #include "Profiler.h"
 
 #if defined(USE_CHRONO)
+#undef __PROFILER_SMP__
 #include <atomic>
 #endif
 

--- a/contrib/profiler/Profiler.cpp
+++ b/contrib/profiler/Profiler.cpp
@@ -23,6 +23,10 @@
 
 #include "Profiler.h"
 
+#if defined(USE_CHRONO)
+#include <atomic>
+#endif
+
 #if defined(__ICC) || defined(__ICL)
 	#pragma warning( disable: 1684 ) // (size_t )name >> 5
 	#pragma warning( disable: 1011 ) // missing return statement at end of non-void function
@@ -59,38 +63,45 @@
 namespace Profiler {
 
 	#if defined(__PROFILER_SMP__)
-		#if defined(_MSC_VER)
+		#if !defined(USE_CHRONO)
+			#if defined(_MSC_VER)
 
-			template< class type >
-			inline bool CAS( volatile type &ptr_, const type old_, const type new_ ) {
-				__asm {
-					mov eax, [old_]
-					mov edx, [new_]
-					mov ecx, [ptr_]
-					lock cmpxchg dword ptr [ecx], edx
-					sete al
+				template< class type >
+				inline bool CAS( volatile type &ptr_, const type old_, const type new_ ) {
+					__asm {
+						mov eax, [old_]
+						mov edx, [new_]
+						mov ecx, [ptr_]
+						lock cmpxchg dword ptr [ecx], edx
+						sete al
+					}
 				}
+
+			#elif defined(__GNUC__) || defined(__ICC)
+
+				template< class type >
+				inline bool CAS( volatile type &ptr_, const type old_, const type new_ ) {
+					u8 ret;
+					__asm__ __volatile__ (
+						"  lock\n"
+						"  cmpxchgl %2,%1\n"
+						"  sete %0\n"
+							: "=q" (ret), "=m" (ptr_)
+							: "r" (new_), "m" (ptr_), "a" (old_)
+							: "memory"
+					);
+					return ret;
+				}
+
+			#else
+				#error Define a compare-and-swap / full memory barrier implementation!
+			#endif
+	#else
+
+			inline bool CAS( volatile std::atomic_uint32_t &ptr_, u32 old_, u32 new_ ) {
+				return std::atomic_compare_exchange_weak(&ptr_, &old_, new_);
 			}
-
-		#elif defined(__GNUC__) || defined(__ICC)
-
-			template< class type >
-			inline bool CAS( volatile type &ptr_, const type old_, const type new_ ) {
-				u8 ret;
-				__asm__ __volatile__ (
-					"  lock\n"
-					"  cmpxchgl %2,%1\n"
-					"  sete %0\n"
-						: "=q" (ret), "=m" (ptr_)
-						: "r" (new_), "m" (ptr_), "a" (old_)
-						: "memory"
-				);
-				return ret;
-			}
-
-		#else
-			#error Define a compare-and-swap / full memory barrier implementation!
-		#endif
+	#endif
 
 		struct CASLock {
 			void Acquire() { while ( !CAS( mLock, u32(0), u32(1) ) ) YIELD() }
@@ -99,7 +110,11 @@ namespace Profiler {
 			bool TryRelease() { return CAS( mLock, u32(1), u32(0) ); }
 			u32 Value() const { return mLock; }
 		//protected:
+			#if !defined(USE_CHRONO)
 			volatile u32 mLock;
+			#else
+			std::atomic_uint32_t mLock;
+			#endif
 		};
 	#else
 		struct CASLock {

--- a/contrib/profiler/Profiler.h
+++ b/contrib/profiler/Profiler.h
@@ -11,7 +11,7 @@
 #undef fastcall
 
 //#define USE_CHRONO
-#if !defined(USE_CHRONO) && defined(__arm__)
+#if !defined(USE_CHRONO) && defined(__arm__) || defined(__aarch64__)
 // this isn't optional for __arm__ builds
 #define USE_CHRONO
 #endif

--- a/contrib/profiler/Profiler.h
+++ b/contrib/profiler/Profiler.h
@@ -11,7 +11,7 @@
 #undef fastcall
 
 //#define USE_CHRONO
-#if !defined(USE_CHRONO) && defined(__arm__) || defined(__aarch64__)
+#if !defined(USE_CHRONO) && (defined(__arm__) || defined(__aarch64__))
 // this isn't optional for __arm__ builds
 #define USE_CHRONO
 #endif

